Add caret and vcaret

benkiel opened the pull request:
As per discussion here https://github.com/unified-font-object/ufo-spec/issues/89

benkiel committed:
Add hcaret and caret

justvanrossum commented on versions/ufo3/glyphs/glif.md:
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
+
+##### hcaret and vcaret
+
+Anchors with a `name` starting with `hcaret` or `vcaret` indicate horizontal or vertical position for the [GDEF Ligature Caret List table]. Anchors with the name `hcaret` provide the `x` value for a ligature caret, and the `y` value is ignored. Likewise, anchors with the name `ycaret` provide the `y` values for a ligature caret, and the `x` value is ignored.
```
I would like the coupling to OpenType GDEF to be looser. In general UFO is export-format agnostic. Maybe mention GDEF caret stuff as an "for example".

Something like
"Anchors with a `name` starting with `hcaret` or `vcaret` indicate horizontal or vertical caret positions for ligature glyphs, for example to compile the OpenType GDEF Ligature Caret List table."

justvanrossum commented on versions/ufo3/glyphs/glif.md:
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
+Both names may be followed by an underscore and arbitrary text (`_X`) for designers to have unique caret position names. The order of carets written to the GDEF table is based upon ascending value order, the `_X` text is not taken into account.
+
+`hcaret` and `vcaret` must not both occur in a glyph.
+
```
Maybe mention that it is expected to have N-1 caret positions for N ligature components.

benkiel committed:
Decouple UFO from OpenType spec and clarify number of caret anchors

There has to be at least one less caret anchor than ligature components (I guess one may want to skip over more than one glyph, so n-1 isn't a must)

moyogo commented:
I now think the distinction between horizontal and vertical carets is not necessary. The glyphs direction should indicate what type of caret it is. It can be just named `caret` instead of `hcaret` or `vcaret`.

anthrotype commented:
in theory (though unlikely) the same glyph could potentially be mapped to different unicodes whose script have conflicting directions.... but we probably should not support this limit case.

benkiel commented:
@moyogo would it be helpful to have the distinction for cases where the direction isn't clear? I know the h/v is fussier but it is explicitly clear.

justvanrossum commented:
I would prefer to be explicit here instead of depending on the compiler having to determine verticality based on the unicode. May save us some epic explanations, though :)

anthrotype commented:
have we thought about using guidelines instead of anchors? I believe guidelines can have a name assigned. We just require that a glyph only has either all vertical or all horizontal "caret[_X]?" guidlines. Haven't fully thought this through.

typoman commented:
Guidelines make more sense as it's also visually easier to see the caret in the glyph view too.

benkiel commented:
@anthrotype the one wrinkle with guidelines would be that there are glyph and font/layer level guidelines. That's just saying "only glyph level guidelines", but I don't know if that would trip things up in Glyphs or not

anthrotype commented:
Glyphs has its own way to specify and generate this GDEF ligature carets. Even the current proposal would need some work on their side in order to support it. They use the name "caret_N" where N is a digit starting from 1. I don't think they accept just caret, let alone hcaret.
Guidelines seem to me more natural for displaying a caret position, but I don't have a strong opinion on that.

benkiel commented:
Based on that, that's an argument for guideline and just `caret[_X]`. I don't feel strongly that this needs to be as anchors, @justvanrossum?

justvanrossum commented:
Guidelines are drawing aids, and I don't thing they've ever been used to declare data in this way. Visually, I understand the point, but structurally in UFO it feels wrong. But that's about as solid as my argument gets I'm afraid.

justvanrossum commented:
> Glyphs has its own way to specify and generate this GDEF ligature carets. Even the current proposal would need some work on their side in order to support it. They use the name "caret_N" where N is a digit starting from 1. I don't think they accept just caret, let alone hcaret.

Do they use anchors this way? And do they support vertical ligature caret positions? If so, how? Unicode magic?

anthrotype commented:
> do they support vertical ligature caret positions?

I don't think so. @schriftgestalt?

justvanrossum commented:
I can also see us opting for `caret_N` and `vcaret_N` if that will result in cheap interoperability with Glyphs for horizontal carets.

schriftgestalt commented:
>> do they support vertical ligature caret positions?

>I don't think so. @schriftgestalt?

Does anyone has a sample font?

justvanrossum commented:
@schriftgestalt just to make sure: does Glyphs indeed use anchors named `caret_1`, `caret_2` etc. to compile GDEF Ligature Caret Lists?

And if so: is the number significant? Is it an error to have them in arbitrary order? For example from left to right: `caret_2`, `caret_1`, `caret_3`?

schriftgestalt commented:
The number doesn't seem do be significant. Glyphs writes the number sorted by anchor name but makeOTF is sorting the values.

justvanrossum commented:
One more thing, can you confirm whether the N in `caret_N` even needs to be a number? I see that glyphsLib only looks for `caret_` and ignores everything that comes after.

benkiel committed:
Change to caret and relax rule limiting horizontal & vertical not being in one glyph

Changed to caret to allow better interoperability with Glyphs. There isn't a reason to limit a glyph to having just a horizontal or vertical caret—chances are needing to do this will be rare, but no need for an arbitrary limit here.

justvanrossum commented on versions/ufo3/glyphs/glif.md:
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
+
+##### Ligature hcaret and vcaret
+
+Anchors with a `name` starting with `hcaret` or `vcaret` indicate horizontal or vertical position for ligature glyphs, for example to complie the [OpenType GDEF Ligature Caret List table]. Anchors with the name `hcaret` provide the `x` value for a ligature caret, and the `y` value is ignored. Likewise, anchors with the name `ycaret` provide the `y` values for a ligature caret, and the `x` value i ...
```
- As per the discussion, I would love for `hcaret` to be renamed to `caret`, and that we require the anchor name to start with `caret_` and `vcaret_` respectively, in both occurances.
- "complie" -> "compile"
- "ycaret" -> "vcaret"

justvanrossum commented on versions/ufo3/glyphs/glif.md:
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
+
+Both names may be followed by an underscore and arbitrary text (`_X`) for designers or applications to have unique caret position names.
+
+`hcaret` and `vcaret` must not both occur in a glyph. There must be at least `n`-1 caret postions for `n` ligature components.
```
- hcaret -> caret
- also add underscores here
- "at least" -> "exactly"
- minor: I would perhaps not use backticks around "n" here, but simply use unformatted cap N: "There must be exactly N-1 caret postions for N ligature components."

benkiel replied:
What if someone wanted to skip over a glyph in a ligature? I don't think we should say exactly, but "no more than".

benkiel committed:
Changes based on feedback

benkiel commented:
I took out the requirement that only `caret` or `vcaret` be in a glyph, as it's an arbitrary limit that doesn't need to be there.

schriftgestalt commented:
> One more thing, can you confirm whether the N in caret_N even needs to be a number? I see that glyphsLib only looks for caret_ and ignores everything that comes after.

The number is ignored so anything is fine.

justvanrossum commented on versions/ufo3/glyphs/glif.md:
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
+
+##### Ligature Carets
+
+Anchors with a `name` starting with `caret_` or `vcaret_` indicate horizontal or vertical position for ligature glyphs, for example to compile the [OpenType GDEF Ligature Caret List table]. Anchors with the name `caret_` provide the `x` value for a ligature caret, and the `y` value is ignored. Likewise, anchors with the name `vcaret_` provide the `y` values for a ligature caret, and the `x` value ...
```
"indicate horizontal or vertical position for ligature glyphs" ->
"indicate horizontal or vertical caret positions for ligature glyphs"

justvanrossum commented on versions/ufo3/glyphs/glif.md:
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
+
+Both names may be followed arbitrary text (`caret_1`, `caret_blah`, etc) for designers or applications to have unique caret position names.
+
+There must be no more than N-1 caret postions for N ligature components.
```
I would still say "There must be exactly N-1 caret positions for N ligature components." (also note the typo in positions)

benkiel committed:
Update glif.md

justvanrossum commented on versions/ufo3/glyphs/glif.md:
```diff
@@ -167,6 +167,18 @@ This element may occur any number of times.
+
+##### Ligature Carets
+
+Anchors with a `name` starting with `caret_` or `vcaret_` indicate horizontal or vertical caret positions for ligature glyphs, for example to compile the [OpenType GDEF Ligature Caret List table]. Anchors with the name `caret_` provide the `x` value for a ligature caret, and the `y` value is ignored. Likewise, anchors with the name `vcaret_` provide the `y` values for a ligature caret, and the `x ...
```
Sorry, a few more nitpicky things:
"Anchors with the name `caret_`" -> "Anchors with a name starting with `caret_`"
"value for a ligature caret" -> "value for a ligature caret position"
"anchors with the name `vcaret_`" -> "anchors with a name starting with `vcaret_`"
"provide the `y` values for a ligature caret" -> "provide the `y` value for a ligature caret position"

benkiel committed:
Edits based on feedback

Cleaning and clarifying language.

benkiel committed:
Update glif.md

benkiel commented:
My bad. Fixed the plural anchors.

benkiel committed:
Clean up language

benkiel committed:
Language cleanup, again

justvanrossum reviewed (approved):
Looks good to me. Please request a review from Tal.

benkiel commented:
@typesupply Can you review and approve?

typesupply commented:
Looks good to me.

benkiel commented:
@typesupply Thank you! Can you merge go ahead and merge, I don't have those rights.

khaledhosny commented:
An idea for future UFO versions; either specify that anchors can have x or y omitted and then they would be displayed as lines instead of points, or define a new structure that would have such property. This can be used for ligature caret as well as MATH italic correction and top accent positions (both are x only), and any potential new property that would need this.
